src/sentry/static/sentry/app/views/settings/projectDataPrivacy/projectDataPrivacyRulesPanel.tsx

```diff
@@ -0,0 +1,214 @@
+import {DATA_TYPE, ACTION_TYPE} from './utils';
+
+type Rule = Omit<
+  React.ComponentProps<typeof ProjectDataPrivacyRulesForm>,
+  'onDelete' | 'onChange'
+>;
+
+type State = {
+  rules: Array<Rule>;
+  savedRules: Array<Rule>;
+  status?: 'success' | 'loading' | 'error';
+};
+
+let requestResponse: Array<Rule> = [
+  {
+    id: 1,
+    action: ACTION_TYPE.MASK,
+    data: DATA_TYPE.BANK_ACCOUNTS,
+    from: 'api_key && !$object',
+  },
+  {
+    id: 2,
+    action: ACTION_TYPE.REMOVE,
+    data: DATA_TYPE.IP_ADDRESSES,
+    from: 'xxx && xxx',
+  },
+];
+
+class ProjectDataPrivacyRulesPanel extends React.Component<{}, State> {
+  state: State = {
+    rules: [],
+    savedRules: [],
+    status: 'loading',
+  };
+
+  componentDidMount() {
+    this.loadRules();
+  }
+
+  loadRules = async () => {
+    // add request here
+    const result: Array<Rule> = await new Promise(resolve => {
+      setTimeout(async function() {
+        resolve(requestResponse);
+      }, 1000);
+    });
+
+    this.setState({
+      rules: result,
+      savedRules: result,
+      status: undefined,
+    });
+  };
+
+  formRef: React.RefObject<HTMLFormElement> = React.createRef();
+
+  handleAddRule = () => {
+    this.setState(prevState => ({
+      rules: [
+        ...prevState.rules,
+        {
+          id: prevState.rules.length + 1,
+          action: ACTION_TYPE.MASK,
+          data: DATA_TYPE.BANK_ACCOUNTS,
+          from: '',
+        },
+      ],
+    }));
+  };
+
+  handleDeleteRule = (ruleId: number) => () => {
+    this.setState(prevState => ({
+      rules: prevState.rules.filter(rule => rule.id !== ruleId),
+    }));
+  };
+
+  handleChange = (updatedRule: Rule) => {
+    this.setState(prevState => ({
+      ...prevState,
+      rules: prevState.rules.map(rule => {
+        if (rule.id === updatedRule.id) {
+          return updatedRule;
+        }
+        return rule;
+      }),
+    }));
+  };
+
+  handleSubmit = () => {
+    this.setState(
+      prevState => ({
+        status: 'success',
+        savedRules: prevState.rules,
+      }),
+      () => {
+        requestResponse = this.state.rules;
+      }
+    );
+  };
+
+  handleValidation = () => {
+    const {rules} = this.state;
+    const isAnyRuleFieldEmpty = rules.find(rule =>
+      Object.keys(rule).find(ruleKey => !rule[ruleKey])
+    );
+
+    const isFormValid = !isAnyRuleFieldEmpty;
+
+    if (isFormValid) {
+      this.formRef.current?.dispatchEvent(new Event('submit'));
+    } else {
+      this.setState({
+        status: 'error',
+      });
+    }
+  };
+
+  handleSaveForm = (event: React.MouseEvent) => {
+    event.stopPropagation();
+    this.handleValidation();
+  };
+
+  handleCancelForm = () => {
+    this.setState(prevState => ({
+      rules: prevState.savedRules,
+    }));
+  };
+
+  render() {
+    const {rules, status} = this.state;
+
+    if (status === 'loading') {
+      addMessage(t('Loading...'), status, {duration: 1000});
```

Comment:
We should not trigger side effects in render

Reply:
Let's use the more specific `addLoadingMessage`, `addSuccessMessage`, and `addErrorMessage`, they have default durations.

Reply:
the main reason I am not using a more specific components is that I had a couple os issues with the time duration + I am already importing addMessage and I have the status (loading, success and error)...so less imports...